src/FwdState.cc

```diff
@@ -722,6 +722,7 @@ FwdState::handleUnregisteredServerEnd()
 void
 FwdState::connectDone(const Comm::ConnectionPointer &conn, Comm::Flag status, int xerrno)
 {
+    n_tries++;
```

Comment:
retryOrBail() does not initiate a new ConnOpener job, which in turn calls FwdState::connectDone(). I believe that all possible results of that job lead to FwdState::connectDone(), including errors and timeouts.

Reply:
> Alex: Why do we want to increment n_tries here if the code below is not guaranteed to use a different destination path?

> Eduard: retryOrBail() does not initiate a new ConnOpener job

I am not sure why you are talking about this here, but, in most cases, it does: retryOrBail() calls startConnectionOrFail() calls connectStart(). One can start to suspect that conclusion without looking at the code because to retry (a failed attempt to open a connection) we may need to open a new connection (naturally).

> I believe that all possible results of a ConnOpener job lead to FwdState::connectDone()

That may be true, but seems irrelevant to my original concern.

I will try to generalize my concern in hope that you focus on the code logic rather than on just making sure that the current code "works" in your tests: Given the currently documented PR intent to tie `n_tries` to the `forward_max_tries` limit on `how many different forward paths Squid will try`, we should increment `n_tries` when Squid uses a new forward path. We need to find the best place(s) to count those uses.

The beginning of a FwdState::connectDone() method is a strange place to count such uses (even if it works):

1. Not all new connections are going to a new address (i.e., we should be concerned about excessive increments).
1. I see no point in waiting for ConnOpener job to finish to count the attempt, even if there is no way for that job to finish without calling connectDone(). Waiting muddies the intent, raises questions, and increases risks that future code modifications will miss a different pathway for that job to finish. Why wait?

Again, please note that this particular concern is not about whether the current PR code "works" in your tests. It is about finding the best place to increment the counter. There may be several such places, but the current choice does not seem like one of them.

Reply:
> I am not sure why you are talking about this here, but, in most cases, it does: retryOrBail() calls startConnectionOrFail() calls connectStart().

Sorry, my statement was inaccurate, it should be that "retryOrBail() does not reuse the same destination's path for the new ConnOpener job".

Reply:
>  when Squid uses a new forward path.

I assume that by start 'using' a new path you mean Squid sending the first request byte. Let's consider SSL peers. When the TCP connection is established, we create BlindPeerConnector() job. Does this SSL negotiation mean that we started using connection? I.e., where should we increment our counter in this case: before SSL negotiation or before sending the first request byte (over established SSL connection)?

Reply:
> I assume that by start 'using' a new path you mean Squid sending the first request byte.

If Squid is reusing an old connection, yes (roughly; see below for more). For _new_ connections, it is the attempt to open it that should be considered destination "use" IMO. Regardless of the actual definition you end up proposing, it should be defined/clarified in forward_max_tries to avoid such questions in the future.

> Let's consider SSL peers. When the TCP connection is established, we create BlindPeerConnector() job. Does this SSL negotiation mean that we started using connection? 

For new connections, IMO, we start using a destination when we call (TCP) connect() or equivalent, which happens even before Squid sends SSL Hello on that connection.

For old/persistent connections, the use attempt should be counted when we write to the first byte to the connection. In SSL case, that first byte may contain no HTTP request info at all.

To simplify, I think it is OK to count uses even earlier -- when Squid _decides_ to open a new connection (to the new destination) and when Squid _decides_ to reuse an old one (to the new destination). Everything that happens after that (including encryption, secondary lookups, certificate requests, etc.) is all irrelevant for forward_max_tries IMO. The forward_max_tries directive is meant to limit the number of those decisions (because they can be expensive to make and to fulfill).

Please note that whether we should honor the "new destination" aspect of the existing forward_max_tries documentation is an open question. My answers above assume the current documentation, but that does not imply that I think that documentation should stay intact. I have not studied that question. You should.


> As far as a single integer counter is concerned, I am afraid that it is not possible to find such, 'best' place, which makes things simple. 

If it is possible to implement forward_max_tries support at all, then such best place(s) do exist -- naturally, out of all the possible working implementations, one or more implementations can be considered the "best" for any given evaluation criteria. Code simplicity should be an evaluation criterion, of course, but it is rarely the only one.

Whether the best PR adjusts forward_max_tries definition and/or uses a single counter are open questions. I have not studied them. You should.